File: blue-core/src/blue/orchestra/GenericInstrument.java
/*
 * blue - object composition environment for csound
 * Copyright (c) 2001-2003 Steven Yi (stevenyi@gmail.com)
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published
 * by  the Free Software Foundation; either version 2 of the License or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.LIB.  If not, write to
 * the Free Software Foundation Inc., 59 Temple Place - Suite 330,
 * Boston, MA  02111-1307 USA
 */

package blue.orchestra;

import blue.Tables;
import blue.udo.OpcodeList;
import blue.utility.TextUtilities;
import blue.utility.UDOUtilities;
import electric.xml.Element;
import electric.xml.Elements;
import java.io.Serializable;
import java.util.HashMap;

/**
 * Title: blue Description: an object composition environment for csound
 * Copyright: Copyright (c) 2001 Company: steven yi music
 * 
 * @author steven yi
 * @version 1.0
 */

public class GenericInstrument extends AbstractInstrument implements
        Serializable {

    String instrumentText;

    String globalOrc;

    String globalSco;

    OpcodeList opcodeList;

    private transient HashMap udoReplacementValues;

    public GenericInstrument() {
        instrumentText = "";
        globalOrc = "";
        globalSco = "";
        opcodeList = new OpcodeList();
    }

    public String generateInstrument() {
        String retVal = instrumentText;

        // the case for auto-generated generic instruments
        if (udoReplacementValues != null) {
            retVal = TextUtilities.replaceOpcodeNames(udoReplacementValues,
                    retVal);
            udoReplacementValues = null;
        }

        return retVal;
    }

    public String getText() {
        return instrumentText;
    }

    public void setText(String iText) {
        if (iText == null) {
            this.instrumentText = "";
        } else {
            this.instrumentText = iText;
        }
    }

    public boolean hasFTable() {
        return false;
    }

    public void generateUserDefinedOpcodes(OpcodeList udoList) {
        udoReplacementValues = UDOUtilities.appendUserDefinedOpcodes(
                opcodeList, udoList);
    }

    public void generateFTables(Tables tables) {
    }

    @Override
    public String toString() {
        return this.name;
    }

    /*
     * (non-Javadoc)
     * 
     * @see blue.orchestra.Instrument#generateGlobalOrc()
     */
    public String generateGlobalOrc() {
        return getGlobalOrc();
    }

    /*
     * (non-Javadoc)
     * 
     * @see blue.orchestra.Instrument#generateGlobalSco()
     */
    public String generateGlobalSco() {
        return getGlobalSco();
    }

    /**
     * @return Returns the globalOrc.
     */
    public String getGlobalOrc() {
        return globalOrc;
    }

    /**
     * @param globalOrc
     *            The globalOrc to set.
     */
    public void setGlobalOrc(String globalOrc) {
        if (globalOrc == null) {
            this.globalOrc = "";
        } else {
            this.globalOrc = globalOrc;
        }
    }

    /**
     * @return Returns the globalSco.
     */
    public String getGlobalSco() {
        return globalSco;
    }

    /**
     * @param globalSco
     *            The globalSco to set.
     */
    public void setGlobalSco(String globalSco) {
        if (globalSco == null) {
            this.globalSco = "";
        } else {
            this.globalSco = globalSco;
        }
    }

    public static Instrument loadFromXML(Element data) throws Exception {
        GenericInstrument instr = new GenericInstrument();

        InstrumentUtilities.initBasicFromXML(data, instr);

        Elements elements = data.getElements();

        while (elements.hasMoreElements()) {
            Element node = elements.next();
            String nodeName = node.getName();

            if (nodeName.equals("globalOrc")) {
                instr.setGlobalOrc(node.getTextString());
            } else if (nodeName.equals("globalSco")) {
                instr.setGlobalSco(node.getTextString());
            } else if (nodeName.equals("instrumentText")) {
                instr.setText(node.getTextString());
            } else if (nodeName.equals("opcodeList")) {
                instr.opcodeList = OpcodeList.loadFromXML(node);
            }
        }

        return instr;
    }

    /*
     * (non-Javadoc)
     * 
     * @see blue.orchestra.Instrument#saveAsXML()
     */
    public Element saveAsXML() {
        Element retVal = InstrumentUtilities.getBasicXML(this);

        retVal.addElement("globalOrc").setText(this.getGlobalOrc());
        retVal.addElement("globalSco").setText(this.getGlobalSco());
        retVal.addElement("instrumentText").setText(this.getText());

        retVal.addElement(opcodeList.saveAsXML());

        return retVal;
    }

    public OpcodeList getOpcodeList() {
        return opcodeList;
    }

}

File: packages/blue-data/src/instruments/generic-instrument.ts
/**
 * GenericInstrument — an instrument defined by raw Csound orchestra text.
 * Mirrors the Java GenericInstrument class.
 */
import { Instrument } from './instrument';
import { Element } from '../serialization/xml-reader';
import { ObjRefSaveMap } from '../serialization/obj-ref-map';
import { DeepCopyable } from '../deep-copyable';
import { OpcodeList } from '../opcodes/opcode-list';
import { appendUserDefinedOpcodes } from '../opcodes/udo-utilities';
import { replaceOpcodeNames } from '../utilities/text';

export class GenericInstrument extends Instrument implements DeepCopyable<GenericInstrument> {
  private _text = '';
  private _globalOrc = '';
  private _globalSco = '';
  private _opcodeList = new OpcodeList();
  private _udoReplacementValues: Map<string, string> | null = null;

  constructor() {
    super();
    this.setName('untitled');
  }

  getText(): string {
    return this._text;
  }

  setText(text: string): void {
    this._text = text ?? '';
  }

  getGlobalOrc(): string {
    return this._globalOrc;
  }

  setGlobalOrc(globalOrc: string): void {
    this._globalOrc = globalOrc ?? '';
  }

  getGlobalSco(): string {
    return this._globalSco;
  }

  setGlobalSco(globalSco: string): void {
    this._globalSco = globalSco ?? '';
  }

  getOpcodeList(): OpcodeList {
    return this._opcodeList;
  }

  setOpcodeList(opcodeList: OpcodeList): void {
    this._opcodeList = opcodeList;
  }

  override generateUserDefinedOpcodes(udoList: unknown): void {
    if (!(udoList instanceof OpcodeList)) {
      return;
    }

    this._udoReplacementValues = appendUserDefinedOpcodes(this._opcodeList, udoList);
  }

  override generateGlobalOrc(): string | null {
    return this._globalOrc || null;
  }

  override generateGlobalSco(): string | null {
    return this._globalSco || null;
  }

  override generateInstrument(): string {
    let instrumentText = this._text;
    if (this._udoReplacementValues && this._udoReplacementValues.size > 0) {
      instrumentText = replaceOpcodeNames(this._udoReplacementValues, instrumentText);
      this._udoReplacementValues = null;
    }

    return instrumentText;
  }

  // ─── XML ───

  saveAsXML(_objRefMap?: ObjRefSaveMap): Element {
    const elem = new Element('instrument');
    elem.setAttribute('type', 'blue.orchestra.GenericInstrument');
    elem.setAttribute('enabled', this._enabled.toString());
    elem.addElement('name').setText(this._name);
    elem.addElement('comment').setText(this._comment);
    elem.addElement('globalOrc').setText(this._globalOrc);
    elem.addElement('globalSco').setText(this._globalSco);
    elem.addElement('instrumentText').setText(this._text);
    elem.addElement(this._opcodeList.saveAsXML());
    return elem;
  }

  static loadFromXML(data: Element): GenericInstrument {
    const instr = new GenericInstrument();
    const name = data.getTextString('name');
    if (name !== null) {
      instr.setName(name);
    }
    instr.setComment(data.getTextString('comment') ?? '');
    instr.setEnabled(data.getAttribute('enabled') !== 'false');
    instr.setText(data.getTextString('instrumentText') ?? '');
    const go = data.getTextString('globalOrc');
    if (go !== null) instr._globalOrc = go;
    const gs = data.getTextString('globalSco');
    if (gs !== null) instr._globalSco = gs;
    const opcodeList = data.getElement('opcodeList');
    if (opcodeList) instr._opcodeList = OpcodeList.loadFromXML(opcodeList);
    return instr;
  }

  deepCopy(): GenericInstrument {
    const copy = new GenericInstrument();
    copy._name = this._name;
    copy._enabled = this._enabled;
    copy._comment = this._comment;
    copy._text = this._text;
    copy._globalOrc = this._globalOrc;
    copy._globalSco = this._globalSco;
    copy._opcodeList = OpcodeList.loadFromXML(this._opcodeList.saveAsXML());
    return copy;
  }
}
